snappy-internal.h

```diff
@@ -250,6 +248,14 @@ static inline std::pair<size_t, bool> FindMatchLength(const char* s1,
           : "+r"(a2)
           : "r"(a3), "r"(xorval)
           : "cc");
+#elif defined(__aarch64__)
+      asm("cmp %w[xorval], 0\n\t"
+          "csel %x[a2], %[a3], %[a2], eq\n\t"
+          : [a2] "+r" (a2)
+          : [a3] "r" (a3) , [xorval] "r" (xorval)
+          : "cc");
```

Comment:
I'm not sure if this is what clang was generating for the default code on aarch64, but this the direct translation from the x86 asm. I verified that it is faster.

Reply:
What does the generated code look before/after? Both GCC and LLVM should use CSEL for this, so I don't understand why this would help. Please file bug reports if not, otherwise it will never get fixed!

Note using random asm like this is generally a bad idea. If absolutely needed, it would be best to abstract it in a generic macro or inline function.

Reply:
> Note using random asm like this is generally a bad idea. If absolutely needed, it would be best to abstract it in a generic macro or inline function.

I agree! However, this code was _already_ using inline asm like this right here for x86, and I am trying to follow local style. All I'm doing is adding a direct translation of the x86 asm to aarch64.

> Both GCC and LLVM should use CSEL for this, so I don't understand why this would help. Please file bug reports if not, otherwise it will never get fixed!

The code already has a comment right above in the x86 case describing why clang doesn't use cmov: https://github.com/google/snappy/blob/5ec5d16bd6896a8ca522a25b19d070e75ab9dd42/snappy-internal.h#L237-L245 On a hunch that the same reasoning also applied to gcc and aarch64, I tried doing the direct port and observed improved perf. Here is the compression perf without this change (since this isn't called in the decompression path):

```
BM_ZFlat/0                      105175 ns       105174 ns         6539 bytes_per_second=928.519M/s html (22.24 %)
BM_ZFlat/1                     1385523 ns      1385503 ns          504 bytes_per_second=483.263M/s urls (47.84 %)
BM_ZFlat/2                        6700 ns         6699 ns       104479 bytes_per_second=17.1121G/s jpg (99.95 %)
BM_ZFlat/3                         333 ns          333 ns      2101538 bytes_per_second=572.801M/s jpg_200 (73.00 %)
BM_ZFlat/4                       11754 ns        11754 ns        58783 bytes_per_second=8.11358G/s pdf (83.31 %)
BM_ZFlat/5                      453779 ns       453780 ns         1540 bytes_per_second=860.825M/s html4 (22.52 %)
BM_ZFlat/6                      430039 ns       430031 ns         1623 bytes_per_second=337.286M/s txt1 (57.87 %)
BM_ZFlat/7                      374322 ns       374315 ns         1865 bytes_per_second=318.929M/s txt2 (62.02 %)
BM_ZFlat/8                     1159839 ns      1159756 ns          601 bytes_per_second=350.923M/s txt3 (55.17 %)
BM_ZFlat/9                     1523393 ns      1523306 ns          459 bytes_per_second=301.672M/s txt4 (66.41 %)
BM_ZFlat/10                      97944 ns        97940 ns         6771 bytes_per_second=1.12766G/s pb (19.61 %)
BM_ZFlat/11                     370635 ns       370624 ns         1881 bytes_per_second=474.284M/s gaviota (37.73 %)
BM_ZFlatAll                    6184921 ns      6184674 ns          111 bytes_per_second=451.585M/s 12 kTestDataFiles
```

The full results with the change are in the PR description, but to summarize, `BM_ZFlatAll` shows 473.433M/s, a 4.8% improvement. The `pb` case (protobufs?) was the most improved at 16%.

Based on that comment, I'm not sure that a bug report would necessarily be helpful here. It appears that while compilers are able to generate csel/cmov here, they currently decide not to because they have reason to suspect that in the general case it would be more harmful than helpful. However, in this specific case, there is local knowledge that it is worth using cmov/csel here, and with a fairly large end-to-end impact. Extending from what is said in the comment, (some with my own wild guesses) one of the issues is that the desired code here adds an unconditional dependency on a load that the compiler would otherwise move inside the branch and therefore could be avoided entirely. In this case it is very very likely to already be in L1, but the compiler may not know that, and if it wasn't that would be very bad for perf. If that is part of the reason, then it seems totally reasonable for the compiler to be conservative about adding data dependencies to loads.

That said, I *do* plan to file some gcc bugs after this. In particular, the codegen for memmove-by-memcpy-through-temp-array is [really bad](https://www.godbolt.org/z/rxMnPc1cP) and has even regressed from the last release.

> What does the generated code look before/after?

It is a bit hard to compare because it this function is inlined into a larger one, and this change changes register allocations around and renumbers labels, but here is the relevant bit (with a few common instructions before and after):

```asm
        rbit    x6, x1
        clz     x6, x6
        asr     w2, w6, 3
        cbnz    w1, .L79
        ldr     x5, [x9, 8]
.L79:
        add     w2, w2, 4
        and     w6, w6, 24
        sub     x4, x9, x4
        sxtw    x2, w2
        add     x1, x9, x2
        lsr     x6, x5, x6
.L80:
        lsl     w5, w4, 8
```
```asm
        rbit    x1, x8
        clz     x1, x1
        and     w7, w1, 24
        lsr     x1, x1, 3
        add     x1, x1, 4
        ldr     x14, [x4, 8]
        add     x4, x4, x1
        cmp w8, 0
        csel x2, x14, x2, eq
        lsr     x7, x2, x7
.L79:
        lsl     w2, w11, 8
```

Reply:
Getting back to this, GCC14 fixes the memmove-by-memcpy-through-temp-array issue (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113618): https://www.godbolt.org/z/qMraefr1K

There is a redundant stack reservation for larger sizes but this won't be an issue in non-trivial functions.